### PR TITLE
Add fasta_nucleotide_color_plot tool

### PR DESCRIPTION
**Pull request by gregvonkuster**

### CODE REVIEW

**Comment by bgruening** [comment]
@gregvonkuster the JAR file is missing.

File: tools/four_color_plot/four_color_plot.xml [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                -N "$n_select_color_cond.n_rgb"
+            #else
+                -N $n_select_color_cond.color_n
+            #end if
+            -o "$output" 1>/dev/null
+        ]]>
+    </command>
+    <inputs>
+        <param name="input" type="data" format="fasta" label="FASTA file" />
+        <param name="pixel_height" type="integer" value="2" min="0" label="Pixel height per nucleotide" />
```
Comment (bgruening) [inline tools/four_color_plot/four_color_plot.xml]: What happens if the pixel hight is really 0? min=1?

File: tools/four_color_plot/four_color_plot.xml [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        <param name="pixel_width" type="integer" value="2" min="0" label="Pixel width per nucleotide" />
+        <conditional name="a_select_color_cond">
+            <param name="a_select_color" type="select" label="A nucleotide: specify RGB value or select color?">
+                <option value="a_rgb" selected="True">specify RGB</option>
+                <option value="a_color">select color</option>
```
Comment (bgruening) [inline tools/four_color_plot/four_color_plot.xml]: Greg do you know that Galaxy has a color picker? Have a look at: https://wiki.galaxyproject.org/Admin/Tools/ToolConfigSyntax?highlight=%28color%29

Comment (bgruening) [inline tools/four_color_plot/four_color_plot.xml]: `<param name="feature_color" type="color" label="Default feature color" value="#ff00ff">`

File: tools/four_color_plot/four_color_plot.xml [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                    <expand macro="color_selector" />
+                </param>
+            </when>
+            <when value="n_rgb">
+                <param name="n_rgb" type="text" value="128,128,128" label="RGB" />
+            </when>
+        </conditional>
+    </inputs>
+    <outputs>
+        <data name="output" format="png" />
```
Comment (bgruening) [inline tools/four_color_plot/four_color_plot.xml]: Is there a option for svg in this tool? For publications?

**Comment by bgruening** [comment]
Great tool! I guess the color picker makes it more simple and useable.

File: tools/four_color_plot/four_color_plot.xml (hunk 2 above)
Comment (gregvonkuster) [inline tools/four_color_plot/four_color_plot.xml]: Goo catch!  Thanks!

File: tools/four_color_plot/four_color_plot.xml (hunk 1 above)
Comment (gregvonkuster) [inline tools/four_color_plot/four_color_plot.xml]: Good catch!  Thanks!

File: tools/four_color_plot/four_color_plot.xml (hunk 3 above)
Comment (gregvonkuster) [inline tools/four_color_plot/four_color_plot.xml]: Is support for svg still active?  Does png not work for publications?

Comment (bgruening) [inline tools/four_color_plot/four_color_plot.xml]: It does, but SVG is superior right? I would offer both options with a select format thingy.

File: tools/four_color_plot/four_color_plot.xml (hunk 2 above)
Comment (bgruening) [inline tools/four_color_plot/four_color_plot.xml]: @guerler does the color picker supports RGB tuples as well? I had the impression it does.
@gregvonkuster if it does not, consider the following code:

```
import struct
foo = (254,252,4)
struct.pack('BBB',*foo).encode('hex')
```

Comment (bgruening) [inline tools/four_color_plot/four_color_plot.xml]: Oh, you need the other way around, right?
Then this one:
```
import struct
value = 'aabbcc'
struct.unpack('BBB', rgbstr.decode('hex'))
```

**Comment by bgruening** [comment]
@gregvonkuster check the tests!!!! It's working!
Still ... do you want to try to color picker?

**Comment by bgruening** [comment]
I mean the tests are working, this is exciting and your tool as well :)

**Comment by gregvonkuster** [comment]
@bgruening I'm willing to try the color picker you think it is beneficial.  I'm still unclear how it will improve the tool though.  The only tool I've found that uses it is browse, and it looks different than this tool.  Will the color picker eliminate the use of the select list?

**Comment by bgruening** [comment]
I hope it will replace all conditionals with 5 color pickers. If this is not what you want you can leave it this way. It's your tool :)

My comments are only suggestions and this tool is completely fine with me as it is. If you think color picker does not add any value, let's merge it.

**Comment by gregvonkuster** [comment]
I'll take a look - thanks!

**Comment by bgruening** [comment]
Just one more comment, maybe a different name would be good. `four_color_plot` can be everything - and it uses 5 colors ;)
Maybe `fasta_nucleotide_color_plot` or something like that?

**Comment by gregvonkuster** [comment]
Do you have examples of how this color picker works?  Here is what JBrowse does, but it is not clear to me if this is what I should be doing.  I don't have time to try to figure this out, so any help is appreciated.  If it is non-trivial to make this change, I'd rather keep the tool the way it is.

    <param label="@LABEL@" type="color" name="@NAME@" value="@VALUE@" help="@HELP@">
      <sanitizer>
        <valid initial="string.letters,string.digits">
          <add value="#" />
        </valid>
      </sanitizer>

**Comment by bgruening** [comment]
I thought it is as easy as:  https://wiki.galaxyproject.org/Admin/Tools/ToolConfigSyntax?highlight=%28color%29

We can also enhance this later if you wish.

**Comment by gregvonkuster** [comment]
I'm back now to looking at this.  Just a sec...

**Comment by gregvonkuster** [comment]
Ok, although the color param is cool, I cannot get it to actually work within the command tag set (I've tried 7 or 8 different combinations and permutations of the logic you advised using struct), so let's just leave it be.  In addition, the tool author wants the ability to enter an RGB value in addition to select a preset color, so using the color param will not eliminate the conditionals.  I like the color param though - just wish I had an example of how to make it work.

I'll rename the tool as you advise though.  Just a sec...

Thanks!

**Comment by bgruening** [comment]
I'm fine with merging and for the improvement maybe @guerler can help.

File: tools/four_color_plot/.shed.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+name: fasta_nucleotide_color_plot
```
Comment (bgruening) [inline tools/four_color_plot/.shed.yml]: Also please the folder to reflect the new name.

File: tools/four_color_plot/.shed.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+owner: iuc
+description: Contains a tool that produces a graphical representation of FASTA data with each nucleotide represented by a selected color.
```
Comment (bgruening) [inline tools/four_color_plot/.shed.yml]: remote_repository_url: is missing

**Comment by bgruening** [comment]
TS's updated.

**Comment by gregvonkuster** [comment]
Thanks @bgruening !!  ;)

**Comment by erasche** [comment]
(Day late and a dollar short, sorry @gregvonkuster)

Sorry, the JBrowse use of `color` is complex, because the jbrowse tool is stupidly complex.

It ideally should be as simple as 

```xml
<param name="feature_color" type="color" label="Default feature color" value="#ff00ff">
</param>
```

though I add a sanitizer so my code doesn't have to deal with `__pd__` in my code.

```xml
<param name="feature_color" type="color" label="Default feature color" value="#ff00ff">
  <sanitizer>
    <valid initial="string.letters,string.digits">
      <add value="#" />
    </valid>
  </sanitizer>
</param>
```

you should be able to access `$feature_color = "#ff00ff"` in your cheetah code.

If the tool author is requesting RGB entry, the color parameter (as it is now) may not be appropriate.

That said, personally I'm fond of restricting colour choices, because people are terrible at choosing colour schemes. Internally, in my org's tools, I removed any choices the users had because they kept making ugly/awful things. Now all of my tools are hardcoded to use brewer colour schemes.

As for your particular implementation, I might've suggested something like:

```python
-A "$a_select_color.color_value"
-C "$c_select_color.color_value"
```

```xml
<expand macro="color_selector" nt="A" nt_s="a" default="255,25,24" />
<expand macro="color_selector" nt="C" nt_s="c" default="25,255,24" />
...
<macros>
  <xml name="color_selector" token_nt="N" token_nt_s="n" token_default="0,0,0">
    <conditional name="@NT_S@_select_color_cond">
      <param name="@NT_@_select_color" type="select" label="A nucleotide: specify RGB value or select color?">
        <option value="rgb" selected="True">specify RGB</option>
        <option value="color">select color</option>
      </param>
      <when value="color">
        <param name="color_value" type="select" label="Color" type="color">
        ... sanitizer
      </when>
      <when value="rgb">
        <!-- I don't suppose you could teach them hex? You could simplify your code immensely here by forcing them to use hexidecimal, with a validator -->
        <param name="color_value" type="text" value="@DEFAULT@" label="RGB" />
      </when>
    </conditional>
  </xml>
</macros>
```

**Comment by erasche** [comment]
parameterized macros are really, really awesome. This is probably one of the simplest examples in the JBrowse tool:

https://github.com/galaxyproject/tools-iuc/blob/master/tools/jbrowse/macros.xml#L196 Just define `token_xyz` attributes on your `<macro />` block, then use `@XYZ@` anywhere in your XML

**Comment by gregvonkuster** [comment]
Thanks for the insight Eric.  I wasn’t quite sure what your sanitizer was doing.  It would have been helpful to have more documentation for the color param on the tool config syntax wiki - I probably would have been able to use it with some context like you’ve provided.

Greg

**Comment by erasche** [comment]
@gregvonkuster I'll go ahead and add that to the wiki, the tool config syntax page is pretty awful sometimes.

Maybe a feature to try out if you ever update the tool :)

**Comment by erasche** [comment]
Done 

- https://wiki.galaxyproject.org/Admin/Tools/ToolConfigSyntax#type.3D.22color.22
- https://wiki.galaxyproject.org/Admin/Tools/ToolConfigSyntax#Parameterizing_XML_Macros.2C_2

**Comment by gregvonkuster** [comment]
Thanks @erasche ;)

**Pull request merged**